src/game.cpp

```diff
@@ -11602,14 +11602,46 @@ void game::read()
     auto filter = []( const item &it ) {
         return it.is_book();
     };
-    int pos = inv_for_filter( _("Read:"), filter );
 
-    if (pos == INT_MIN) {
-        add_msg(_("Never mind."));
+    auto item_loc = inv_map_splice( filter, _("Read:") );
+
+    const int inv_pos = item_loc.get_inventory_position();
+
+    if ( inv_pos != INT_MIN ) {
+        draw();
+        u.read( inv_pos, FROM_INVENTORY );
+        return;
+    }
+
+    item *it = item_loc.get_item();
+    if ( it == nullptr ) {
+        add_msg( _("Never mind.") );
+        return;
+    }
+
+    if ( !u.can_pickup(true) ) {
```

Comment:
If pickup is not allowed, books outside inventory shouldn't be listed. Currently it will fail silently if you have bionic claws and try to read a book outside your inventory.